harvester/terraform-provider-harvester
# feat: add harvester_cloudinit_secret

Issue: https://github.com/harvester/harvester/issues/2958

## Test plan

**Setup**
Follow [PR Test Env Setup](https://github.com/harvester/terraform-provider-harvester/wiki/PR-Test-Env-Setup)

**Case 1: Apply same terraform doesn't trigger change**
1. Create a file `main.tf`.
```
resource "harvester_cloudinit_secret" "ex1" {
  name             = "ex1"
  namespace        = "default"
  user_data_base64 = "I2Nsb3VkLWNvbmZpZwpwYWNrYWdlX3VwZGF0ZTogdHJ1ZQpwYWNrYWdlczoKICAtIHFlbXUtZ3Vlc3QtYWdlbnQKcnVuY21kOgogIC0gLSBzeXN0ZW1jdGwKICAgIC0gZW5hYmxlCiAgICAtIC0tbm93CiAgICAtIHFlbXUtZ3Vlc3QtYWdlbnQuc2VydmljZQpwYXNzd29yZDogdGVzdApjaHBhc3N3ZDoKICBleHBpcmU6IGZhbHNlCnNzaF9wd2F1dGg6IHRydWUK"
}

resource "harvester_cloudinit_secret" "ex2" {
    name = "ex2"
    namespace = "default"

    user_data    = <<-EOF
    #cloud-config
    password: test
    chpasswd:
      expire: false
    ssh_pwauth: true
    package_update: true
    packages:
      - qemu-guest-agent
    runcmd:
      - - systemctl
        - enable
        - '--now'
        - qemu-guest-agent
    EOF
    network_data = ""
}
```
2. Run `terraform apply`. Terraform should add `ex1` and `ex2` secrets to the cluster.
3. Run `terraform apply` again. Nothing change.

**Case 2: Change content to empty will update the secret**
1. Follow case 1.
2. Update the file `main.tf`.
```
resource "harvester_cloudinit_secret" "ex1" {
  name             = "ex1"
  namespace        = "default"
  user_data_base64 = ""
}

resource "harvester_cloudinit_secret" "ex2" {
    name = "ex2"
    namespace = "default"

    user_data    = ""
    network_data = ""
}
```
3. Run `terraform apply`. The content of `ex1` and `ex2` secrets will be empty.
4. Run `terraform apply` again. Nothing change.
